## **PROJET DE IMDB REVIEW CLASSIFICATION**

---

In [74]:
from tensorflow.keras.datasets import imdb
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, Flatten, Dense,GlobalMaxPooling1D

### **Importation du dataset**

In [75]:
(X_train,y_train),(X_test,y_test)=imdb.load_data(num_words = 10000)

print(f'Training data shape: {X_train.shape}, Training labels shape: {y_train.shape}')
print(f'Testing data shape: {X_train.shape}, Testing labels shape: {y_test.shape}')

Training data shape: (25000,), Training labels shape: (25000,)
Testing data shape: (25000,), Testing labels shape: (25000,)


Donc nous avons 25000 ligne et un seule colonne soit dans le variable dependate ou bien independate

In [76]:
X_train

array([list([1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 2, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 2, 336, 385, 39, 4, 172, 4536, 1111, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2025, 19, 14, 22, 4, 1920, 4613, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 1247, 4, 22, 17, 515, 17, 12, 16, 626, 18, 2, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2223, 5244, 16, 480, 66, 3785, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 1415, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 2, 8, 4, 107, 117, 5952, 15, 256, 4, 2, 7, 3766, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 2, 1029, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2071, 56, 26, 141, 6, 194, 7486, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 5535, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 1334, 88, 12, 16, 283, 5, 16, 4472, 113, 103, 32, 15, 16, 5345, 19, 178, 32]),
       list([1, 194, 1153, 194, 8255, 78, 228,

In [77]:
## Inspect a sample review and its label
first_ligne = X_train[0]
first_ligne_output = y_train[0]

print(f"first_ligne:{first_ligne}")
print(f'first_ligne_output: {first_ligne_output}')


first_ligne:[1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 2, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 2, 336, 385, 39, 4, 172, 4536, 1111, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2025, 19, 14, 22, 4, 1920, 4613, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 1247, 4, 22, 17, 515, 17, 12, 16, 626, 18, 2, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2223, 5244, 16, 480, 66, 3785, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 1415, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 2, 8, 4, 107, 117, 5952, 15, 256, 4, 2, 7, 3766, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 2, 1029, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2071, 56, 26, 141, 6, 194, 7486, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 5535, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 1334, 88, 12, 16, 283, 5, 16, 4472, 113, 103, 32, 15, 16, 5345, 19, 178, 32]
first_ligne_output: 1


Donc c'est comme nous avons des valeurs encoder(par son index) pour plusieur colonne mais nous avons pas les noms de ces index. mais en peux extraire les valeurs apartir de indexe par exemple pour indexe 1, 16, 530, 1622:

In [78]:
word_index = imdb.get_word_index()

desired_values = {0,1,2,3,4,5}

reverse_word_index = {value: key for key, value in word_index.items() if value in desired_values}
reverse_word_index

{4: 'of', 3: 'a', 1: 'the', 2: 'and', 5: 'to'}

# Modele padding

In [79]:
import numpy as np
print(np.max([np.max(sequence) for sequence in X_train]))


9999


In [87]:
from tensorflow.keras.preprocessing import sequence
from tensorflow.keras.layers import Dropout

max_len = 650 #! C'est la longueur maximale des séquences d'entrée que le modèle prendra en charge si il n'a pas depasser 500 on vas completer avec des 0

X_train=sequence.pad_sequences(X_train,maxlen=max_len)
X_test = sequence.pad_sequences(X_test, maxlen=max_len)

X_test


array([[   0,    0,    0, ...,   14,    6,  717],
       [   0,    0,    0, ...,  125,    4, 3077],
       [   0,    0,    0, ...,    9,   57,  975],
       ...,
       [   0,    0,    0, ...,   21,  846, 5518],
       [   0,    0,    0, ..., 2302,    7,  470],
       [   0,    0,    0, ...,   34, 2005, 2643]])

* Tronquage : Si une critique dépasse 500 mots, elle sera tronquée pour conserver uniquement les 500 premiers mots.

* Complétion : Si une critique contient moins de 500 mots, des zéros seront ajoutés au début ou à la fin pour atteindre la longueur de 500 mots.

### **Creation du modele**

In [88]:
dim = 200  
model = Sequential()

#* input_dim: nombre de mots différents que nous avons dans chaque mot dans les criteres vas representer par un nombre unique dans intervalle [0,10000]
#? input_length : Donc chaque sequence (ligne) sera representer par un vecteur de taille fix = 650 ici ou nous avons avant completer par des 0
#! output_dim : chaque mot(ou bien on peux dire index) parmis les valeurs de sequence de input_length sera trasnformer aussi a un vecteur de len 200

model.add(Embedding(input_dim = 10000, output_dim = dim, input_length = max_len))  # Embedding layer
model.add(GlobalMaxPooling1D()) 
model.add(Dropout(0.5))  
model.add(Dense(1, activation='sigmoid'))


### Exemple: 

Séquence d'indices : [2, 45, 12, 500]

* **input_dim** = 10000 : Il y a un vocabulaire de 10 000 mots.

* **output_dim** = 8 : Chaque mot sera représenté par un vecteur de taille 8.

* **input_length** = 500 : La longueur maximale des séquences.

---

* Indice 2 → Vecteur : [0.1, -0.3, 0.2, 0.7, -0.1, 0.4, 0.0, -0.2]

* Indice 45 → Vecteur : [0.3, -0.5, 0.6, -0.1, 0.3, -0.4, 0.2, 0.1]

* Indice 12 → Vecteur : [0.2, 0.0, -0.2, 0.5, 0.1, -0.3, -0.1, 0.2]

* Indice 500 → Vecteur : [-0.1, 0.2, 0.4, -0.3, 0.0, 0.5, -0.2, 0.3]

---

[[ 0.1, -0.3, 0.2, 0.7, -0.1, 0.4, 0.0, -0.2],

 [ 0.3, -0.5, 0.6, -0.1, 0.3, -0.4, 0.2, 0.1],

 [ 0.2, 0.0, -0.2, 0.5, 0.1, -0.3, -0.1, 0.2],

 [-0.1, 0.2, 0.4, -0.3, 0.0, 0.5, -0.2, 0.3]]


In [89]:
model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

In [90]:
model.build(input_shape=(None, max_len))
model.summary()

Model: "sequential_8"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding_8 (Embedding)         │ (None, 650, 200)       │     2,000,000 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ global_max_pooling1d_1          │ (None, 200)            │             0 │
│ (GlobalMaxPooling1D)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_8 (Dropout)             │ (None, 200)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_8 (Dense)                 │ (None, 1)              │           201 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,000,201 (7.63 MB)

 Trainable params: 2,000,201 (7.63 MB)

 Non-trainable params: 0 (0.00 B)

In [91]:
from tensorflow.keras.callbacks import EarlyStopping
earlystopping=EarlyStopping(monitor='val_loss',patience=8,restore_best_weights=True)
earlystopping

In [92]:
history=model.fit(
    X_train,y_train,epochs=50,batch_size=32,
    validation_split=0.3,
    callbacks=[earlystopping]
)

Epoch 1/50
547/547 ━━━━━━━━━━━━━━━━━━━━ 20s 36ms/step - accuracy: 0.6555 - loss: 0.6394 - val_accuracy: 0.8451 - val_loss: 0.4086
Epoch 2/50
547/547 ━━━━━━━━━━━━━━━━━━━━ 18s 32ms/step - accuracy: 0.8423 - loss: 0.3907 - val_accuracy: 0.8724 - val_loss: 0.3210
Epoch 3/50
547/547 ━━━━━━━━━━━━━━━━━━━━ 18s 32ms/step - accuracy: 0.8790 - loss: 0.3048 - val_accuracy: 0.8812 - val_loss: 0.2906
Epoch 4/50
547/547 ━━━━━━━━━━━━━━━━━━━━ 18s 33ms/step - accuracy: 0.8930 - loss: 0.2654 - val_accuracy: 0.8877 - val_loss: 0.2752
Epoch 5/50
547/547 ━━━━━━━━━━━━━━━━━━━━ 19s 35ms/step - accuracy: 0.9105 - loss: 0.2398 - val_accuracy: 0.8879 - val_loss: 0.2667
Epoch 6/50
547/547 ━━━━━━━━━━━━━━━━━━━━ 24s 44ms/step - accuracy: 0.9210 - loss: 0.2074 - val_accuracy: 0.8936 - val_loss: 0.2629
Epoch 7/50
547/547 ━━━━━━━━━━━━━━━━━━━━ 21s 38ms/step - accuracy: 0.9297 - loss: 0.1819 - val_accuracy: 0.8945 - val_loss: 0.2622
Epoch 8/50
547/547 ━━━━━━━━━━━━━━━━━━━━ 22s 40ms/step - accuracy: 0.9401 - loss: 0.1655 - 

In [93]:
## Save model file
model.save('simple_rnn_imdb.h5')